### DISCUSSION TITLE
How to print traceback of last unhandled exception when reaching Micropython prompt

### DISCUSSION THREAD

**Discussion by ondrej1024** [micropython/micropython#13593] [Using MicroPython]
Suppose my Micropython program stops because it hits an unhandled exception.  How can I print the traceback of this exception from the Micropython prompt ?

I found `sys.print_exception()` but it takes the exception value as argument and `sys.last_exc` which should contain tha last exception value is not available in Micropython.

Thanks, Ondrej

**Comment by kjm1102** [comment]
Why not put the sys.print_exception in your program?
```
try:
 your program
except Exception as e: sys.print_exception(e)
```

**Comment by ondrej1024** [comment]
If I have the Python REPL open during program execution and an unhandled exception occurs, the traceback will already be printed on the console so I know exactly where in the code the exception occurred. But if the exception happens without the REPL open, then the traceback printout will be lost. So I am looking for a way to access the traceback opening the REPL _after_ the exception has occured and the program has stopped.
The exception I am trying to debug is not easily reproducible and could happen after days of running the program.

**Reply by GitHubsSilverBullet** [reply]
Store the exceptions in a global variable (or a small array/list).
Print that one should the need arise.

**Reply by ondrej1024** [reply]
I am talking about an unhandled exception that could occur anywhere in the code, outside a try/except statement. How can I store this kind of exception in a variable?

**Reply by karfas** [reply]
Just **handle** the exception. Nothing prevents you from wrapping your entire program in a try/except block.

**Reply by ondrej1024** [reply]
So if I understand this correctly, there is no possibility of a post mortem debug in Micropython? No way of accessing the last exception from he REPL (unless explicitly handled in the code) ?

**Reply by GitHubsSilverBullet** [reply]
wrap your entire application in a catch-all exception and save it's outcome.

**Comment by karfas** [comment]
> So if I understand this correctly, there is no possibility of a post mortem debug in Micropython? 

What - exactly - is your specific problem with a catch-all exception handler ?

**If you want to see the exception** (without constantly monitoring REPL output), **catch it** and save it somewhere (global variable, file, wherever) and (maybe) print it in parallel.

In my opinion, the difference to other environments is minimal in this regard.  If you want monitoring of your services, you will most likely need to add instrumentation to your code (e.g. logfiles, print to stdout/stderr, ...). Larger systems might give you *some* post-mortem analyzation possibilites (e.g. core dumps, docker logfiles, ...). Do you really expect this kind of features on a microcontroller with a few 100kB of memory ?

**Reply by ondrej1024** [reply]
As I wrote in my initial post, `sys.print_exception()` actually  does (almost) everything I need. I was only wondering why `sys.last_exc` is not available in Micropython as it doesn't seem such a huge resource demanding feature that couldn't be implemented on a microcontroller.

But anyway, as you said, the last exception can be saved by wrapping the whole program in a try/except block, so I will go down this road.

**Comment by shariltumin** [comment]
You can try logging your exceptions to a log file and setting a limit on the number of logs. It is best to write the log to a file on an SDcard, you can then read the card on your PC.

```py
import time, machine, sys

class ERR(Exception):
   def __init__(self, n):
      super().__init__()
      self.n = n
      self.c = 0
   def log(self, error):
      self.c += 1
      if self.c < self.n:
         with open('err.log', 'a+') as f:
             f.write(f'{time.time()}: ')
             sys.print_exception(error, f)
      else:
         machine.soft_reset() # or whatever you need
```

And you can use it like this:
```py
>>> import exc
>>> err = exc.ERR(10)
>>> while 1:
...     try:
...         m = 23/0
...     except Exception as e:
...         err.log(e)
... 
MPY: soft reboot
MicroPython v1.22.0 ....
>>> 
```

**Reply by ondrej1024** [reply]
Thanks for the idea. I will try this, a persistent exception log seems like a very handy tool.
